### ISSUE TITLE
Error when updating 2 different structs within one transaction

### ISSUE THREAD

**Issue by boon-wego** [bitcrowd/carbonite#112]
This is more like a question rather than an issue. Say i have `foo` and `bar` which is related 1 to 1.

```
** (Postgrex.Error) ERROR 42804 (datatype_mismatch) type of parameter 1 (foos) does not match that when preparing the plan (bars)
```
We have some code that updates two tables in a single Multi transaction
```
case Multi.new()
         |> Carbonite.Multi.insert_transaction(%{meta: %{type: "foo_updated"}})
         |> Multi.update(:update_foo, foo_changeset)
         |> Multi.run(:update_bar, fn _repo, _changes ->
           BarAdmin.update(foo.bar_id)
         end)
         |> Repo.transaction() do
```
How are we suppose to deal with such a use case?

**Comment by maltoe** [comment]
Hi @boon-wego,

in what sense do you need to deal with this case? The code you presented looks great. If it is about the `foo_updated` info in the `meta` field: You're free to put whatever you want there, right? This is for you, and for you only, Carbonite doesn't care. In the real world, one could imagine an operation like `customer_checkout` which inserts/updates tables `customers`, `orders`, `invoices`, etc.

**Comment by maltoe** [comment]
Hi @boon-wego 

This error does not appear to be related to Carbonite. If I'd have to guess, I'd say you have a field defined in either your `foo` or your `bar` Ecto schema that doesn't match the table definition.

**Comment by boon-wego** [comment]
This is coming from my test suite. I noticed that if i run individual test files, they pass with no failure. But if i run the whole suite the error happens. I'm adopting this strategy https://github.com/bitcrowd/carbonite?tab=readme-ov-file#option-2-bypassing-change-capture-by-default

Q: When we run tests where we `enable_carbonite`, do we need to make sure these test don't run in parallel

**Comment by boon-wego** [comment]
I managed to write some throw away code to reproduce this problem

```
def foo_create() do
    case Multi.new()
         |> Carbonite.Multi.insert_transaction(%{meta: %{type: "foo_inserted"}})
         |> Multi.run(:create, fn _repo, _changes ->
           %Foo{}
           |> Foo.changeset(%{name: "FOO"})
           |> Repo.insert()
         end)
         |> Multi.run(:create_bar, fn _repo, _changes ->
           bar_create()
         end)
         |> Repo.transaction() do
      {:ok, changes} ->
        {:ok, changes[:create]}

      {:error, _failed_operation, failed_value, _changes_so_far} ->
        {:error, failed_value}
    end
  end

  def bar_create() do
    case Multi.new()
         |> Carbonite.Multi.insert_transaction(%{meta: %{type: "bar_inserted"}})
         |> Multi.run(:create, fn _repo, _changes ->
           %Bar{}
           |> Bar.changeset(%{name: "BAR"})
           |> Repo.insert()
         end)
         |> Repo.transaction() do
      {:ok, changes} ->
        {:ok, changes[:create]}

      {:error, _failed_operation, failed_value, _changes_so_far} ->
        {:error, failed_value}
    end
  end
```
Test
```
describe "temp" do
    test "success" do
      :ok = enable_carbonite()

      assert {:ok, %Foo{} = _deal} = Temp.foo_create()

      carbonite_transaction = current_transaction() 

      assert carbonite_transaction.meta == %{
               "type" => "foo_inserted"
             }
    end
  end
```
Error
`** (Postgrex.Error) ERROR 42804 (datatype_mismatch) type of parameter 1 (bars) does not match that when preparing the plan (foos)`

**Comment by maltoe** [comment]
This is an extended version of what you originally posted, but `Foo` and `Bar` schemas and table definitions are still unknown. As said, the error reads like it has nothing to do with Carbonite. Please provide a full minimum working example (ideally as a `.exs` script, like https://github.com/bitcrowd/carbonite/issues/111#issuecomment-2260602813), if you think otherwise.

**Comment by boon-wego** [comment]
Managed to reproduce with this

.tool-versions
```
elixir 1.15.4-otp-26
erlang 26.0.1
postgres 14.5
```
test.exs
```
# carbonite_test.exs

Mix.install([
  {:postgrex, "> 0.0.0"},
  {:carbonite, "0.14.0"}
])

Application.put_env(:my_app, Repo,
  username: "postgres",
  password: "postgres",
  database: "carbonite_override_test"
)

defmodule Repo do
  use Ecto.Repo,
    otp_app: :my_app,
    adapter: Ecto.Adapters.Postgres
end

defmodule Migration do
  use Ecto.Migration

  def up do
    create table(:foos) do
      add(:name, :string)
    end

    create table(:bars) do
      add(:name, :string)
    end

    Carbonite.Migrations.create_trigger(:foos)
    Carbonite.Migrations.put_trigger_config(:foos, :mode, :capture)
    Carbonite.Migrations.put_trigger_config(:foos, :store_changed_from, true)

    Carbonite.Migrations.create_trigger(:bars)
    Carbonite.Migrations.put_trigger_config(:bars, :mode, :capture)
    Carbonite.Migrations.put_trigger_config(:bars, :store_changed_from, true)
  end

  def down, do: :ok
end

# ecto.create
Repo.__adapter__().storage_up(Repo.config())

# start repo
{:ok, _} = Repo.start_link()

# ecto.migrate
Ecto.Migrator.run(Repo, [{1, Migration}], :up, all: true)

defmodule Foo do
  use Ecto.Schema
  import Ecto.Changeset

  schema "foos" do
    field :name, :string
  end

  def changeset(foo, attrs) do
    foo
    |> cast(attrs, [:name])
  end
end

defmodule Bar do
  use Ecto.Schema
  import Ecto.Changeset

  schema "bars" do
    field :name, :string
  end

  def changeset(bar, attrs) do
    bar
    |> cast(attrs, [:name])
  end
end

defmodule Temp do
  alias Ecto.Multi

  def foo_create() do
    case Multi.new()
         |> Carbonite.Multi.insert_transaction(%{meta: %{type: "foo_inserted"}})
         |> Multi.run(:create, fn _repo, _changes ->
           %Foo{}
           |> Foo.changeset(%{name: "FOO"})
           |> Repo.insert()
         end)
         |> Multi.run(:create_bar, fn _repo, _changes ->
           bar_create()
         end)
         |> Repo.transaction() do
      {:ok, changes} ->
        {:ok, changes[:create]}

      {:error, _failed_operation, failed_value, _changes_so_far} ->
        {:error, failed_value}
    end
  end

  def bar_create() do
    case Multi.new()
         |> Carbonite.Multi.insert_transaction(%{meta: %{type: "bar_inserted"}})
         |> Multi.run(:create, fn _repo, _changes ->
           %Bar{}
           |> Bar.changeset(%{name: "BAR"})
           |> Repo.insert()
         end)
         |> Repo.transaction() do
      {:ok, changes} ->
        {:ok, changes[:create]}

      {:error, _failed_operation, failed_value, _changes_so_far} ->
        {:error, failed_value}
    end
  end
end

Temp.foo_create()
```

**Comment by boon-wego** [comment]
Managed to reproduce with this

.tool-versions
```
elixir 1.15.4-otp-26
erlang 26.0.1
postgres 14.5
```
test.exs
```
Mix.install([
  {:postgrex, "> 0.0.0"},
  {:carbonite, "0.14.0"}
])

Application.put_env(:my_app, Repo,
  username: "postgres",
  password: "postgres",
  database: "carbonite_override_test"
)

defmodule Repo do
  use Ecto.Repo,
    otp_app: :my_app,
    adapter: Ecto.Adapters.Postgres
end

defmodule Migration do
  use Ecto.Migration

  def up do
    create table(:foos) do
      add(:name, :string)
    end

    create table(:bars) do
      add(:name, :string)
    end

    Carbonite.Migrations.create_trigger(:foos)
    Carbonite.Migrations.put_trigger_config(:foos, :mode, :capture)
    Carbonite.Migrations.put_trigger_config(:foos, :store_changed_from, true)

    Carbonite.Migrations.create_trigger(:bars)
    Carbonite.Migrations.put_trigger_config(:bars, :mode, :capture)
    Carbonite.Migrations.put_trigger_config(:bars, :store_changed_from, true)
  end

  def down, do: :ok
end

# ecto.create
Repo.__adapter__().storage_up(Repo.config())

# start repo
{:ok, _} = Repo.start_link()

# ecto.migrate
Ecto.Migrator.run(Repo, [{1, Migration}], :up, all: true)

defmodule Foo do
  use Ecto.Schema
  import Ecto.Changeset

  schema "foos" do
    field :name, :string
  end

  def changeset(foo, attrs) do
    foo
    |> cast(attrs, [:name])
  end
end

defmodule Bar do
  use Ecto.Schema
  import Ecto.Changeset

  schema "bars" do
    field :name, :string
  end

  def changeset(bar, attrs) do
    bar
    |> cast(attrs, [:name])
  end
end

defmodule Temp do
  alias Ecto.Multi

  def foo_create() do
    case Multi.new()
         |> Carbonite.Multi.insert_transaction(%{meta: %{type: "foo_inserted"}})
         |> Multi.run(:create, fn _repo, _changes ->
           %Foo{}
           |> Foo.changeset(%{name: "FOO"})
           |> Repo.insert()
         end)
         |> Multi.run(:create_bar, fn _repo, _changes ->
           bar_create()
         end)
         |> Repo.transaction() do
      {:ok, changes} ->
        {:ok, changes[:create]}

      {:error, _failed_operation, failed_value, _changes_so_far} ->
        {:error, failed_value}
    end
  end

  def bar_create() do
    case Multi.new()
         |> Carbonite.Multi.insert_transaction(%{meta: %{type: "bar_inserted"}})
         |> Multi.run(:create, fn _repo, _changes ->
           %Bar{}
           |> Bar.changeset(%{name: "BAR"})
           |> Repo.insert()
         end)
         |> Repo.transaction() do
      {:ok, changes} ->
        {:ok, changes[:create]}

      {:error, _failed_operation, failed_value, _changes_so_far} ->
        {:error, failed_value}
    end
  end
end

Temp.foo_create()
```
console output
```
elixir test.exs  

09:43:05.169 [info] Migrations already up

09:43:05.220 [debug] QUERY OK db=0.2ms idle=58.9ms
begin []

09:43:05.240 [debug] QUERY OK source="transactions" db=7.3ms
INSERT INTO "carbonite_default"."transactions" ("meta","inserted_at") VALUES ($1,$2) ON CONFLICT ("id") DO UPDATE SET "id" = EXCLUDED."id" RETURNING "inserted_at","meta","xact_id","id" [%{type: "foo_inserted"}, ~U[2024-08-21 01:43:05.229242Z]]

09:43:05.244 [debug] QUERY OK source="foos" db=2.2ms
INSERT INTO "foos" ("name") VALUES ($1) RETURNING "id" ["FOO"]

09:43:05.244 [debug] QUERY OK source="transactions" db=0.1ms
INSERT INTO "carbonite_default"."transactions" ("meta","inserted_at") VALUES ($1,$2) ON CONFLICT ("id") DO UPDATE SET "id" = EXCLUDED."id" RETURNING "inserted_at","meta","xact_id","id" [%{type: "bar_inserted"}, ~U[2024-08-21 01:43:05.244230Z]]

09:43:05.246 [debug] QUERY ERROR source="bars" db=2.2ms
INSERT INTO "bars" ("name") VALUES ($1) RETURNING "id" ["BAR"]

09:43:05.247 [debug] QUERY OK db=0.1ms
rollback []
** (Postgrex.Error) ERROR 42804 (datatype_mismatch) type of parameter 1 (bars) does not match that when preparing the plan (foos)
    (ecto_sql 3.12.0) lib/ecto/adapters/sql.ex:1078: Ecto.Adapters.SQL.raise_sql_call_error/1
    (ecto 3.12.1) lib/ecto/repo/schema.ex:834: Ecto.Repo.Schema.apply/4
    (ecto 3.12.1) lib/ecto/repo/schema.ex:415: anonymous fn/15 in Ecto.Repo.Schema.do_insert/4
    (ecto 3.12.1) lib/ecto/multi.ex:897: Ecto.Multi.apply_operation/5
    (elixir 1.15.4) lib/enum.ex:2510: Enum."-reduce/3-lists^foldl/2-0-"/3
    (ecto 3.12.1) lib/ecto/multi.ex:870: anonymous fn/5 in Ecto.Multi.apply_operations/5
    (ecto_sql 3.12.0) lib/ecto/adapters/sql.ex:1382: anonymous fn/3 in Ecto.Adapters.SQL.checkout_or_transaction/4
    (db_connection 2.7.0) lib/db_connection.ex:1049: DBConnection.transaction/3

```

**Comment by maltoe** [comment]
Thank you.

Reduced version:

```elixir
Mix.install([
  {:postgrex, "> 0.0.0"},
  {:carbonite, "0.14.0"}
])

Application.put_env(:my_app, Repo,
  username: "postgres",
  password: "postgres",
  database: "carbonite_boon_test"
)
defmodule Repo do
  use Ecto.Repo,
    otp_app: :my_app,
    adapter: Ecto.Adapters.Postgres
end

defmodule Migration do
  use Ecto.Migration

  def up do
    create table(:table1) do
    end

    create table(:table2) do
    end

    Carbonite.Migrations.up(1..8)
    Carbonite.Migrations.create_trigger(:table1)
    Carbonite.Migrations.create_trigger(:table2)
  end

  def down, do: :ok
end

# ecto.create
Repo.__adapter__().storage_up(Repo.config())

# start repo
{:ok, _} = Repo.start_link()

# ecto.migrate
Ecto.Migrator.run(Repo, [{1, Migration}], :up, all: true)

defmodule Test do
  def run() do
    Repo.transaction(fn ->
      Carbonite.insert_transaction(Repo)
      Repo.insert_all("table1", [%{}])
      Repo.insert_all("table2", [%{}])
    end)
  end
end

Test.run()
```

Confirmed this happens for Postgres <15 and Carbonite >= 0.12, i.e. it was introduced in migration patch [v8](https://github.com/bitcrowd/carbonite/blob/main/lib/carbonite/migrations/v8.ex).

@boon-wego I'll need to look into this further, but am on vacation currently with limited access to screens. Any chance you can update to Postgres 15 or above?

**Comment by boon-wego** [comment]
No problem, just want to confirm that its got something to do with the carbonite library and not my current codebase. I'm not in a hurry to upgrade.

**Comment by maltoe** [comment]
Yes, this use-case is totally fine and bitcrowd has similar multiple insert transactions in a lot of places in our own projects. Certainly a bug in Carbonite when used with PG <15.

**Comment by maltoe** [comment]
@boon-wego Thanks for reporting this. I was torn between fixing this and just marking Carbonite as "incompatible w/ PG 14.[0-5]" as it's such a weird "bug". I don't even think it could be called a bug in PG, more like a feature they added in 14 and improved to general usefulness in 14.5. When I used it in my refactoring in Carbonite v0.12 (working on PG 16 at the time), it was of course already pretty stable. Ended up fixing it as my "refactorings" didn't improve the intended issue much (lots of duplication in the migration files).

Released as v0.14.2
